normative-conventions.md

```diff
@@ -0,0 +1,37 @@
+# Normative conventions
+
+This document describes some of the conventions that we try to follow when designing new features. Sometimes old features don't follow these; that's ok, new features generally should anyway.
+
+None of these rules are inviolable, but you should have a good reason for any particular feature to deviate (for example, because the feature is a close cousin of an existing feature, like `findLastIndex` vs `findIndex`).
+
+This list is very far from being complete.
+
+## Number-taking inputs should reject `NaN`
+
+In any built-in function which expects a number (including as an option in an options bag), receiving `NaN` or anything which results in `NaN` after coercion is performed (such as by passing through `ToNumeric`, `ToPrimitive`, `ToNumber`, etc.) should cause a `RangeError` to be thrown.
```

Comment:
This is worded too strongly and not something I'd agreed to. I thought we were only discussing functions for which NaN does not make sense as an input.

This rule should apply only to functions for which NaN is not part of the domain. There exist plenty of built-in number-taking functions (pretty much everything in Math, for example) for which NaN is a perfectly fine member of the domain and produces defined behavior. We must not encourage throwing on those.

Reply:
This spec also raises another open question about what happens if NaN is passed to a numeric argument but that argument is not used. Should the function throw or not?

An example might be a function that takes a long list of array indices and returns the first index that satisfies some property. Should it go out of its way to make sure that none of the subsequent indices are NaN just so it can satisfy this rule?

Reply:
Yes, imo extensive validation should be performed on all inputs even if they end up being unused.

Reply:
> This is worded too strongly and not something I'd agreed to. I thought we were only discussing functions for which NaN does not make sense as an input.

Ah, I'd intended that to be covered by "expects to receive a number" (since NaN is not a number), but I see how that wording is confusing. I've updated it to make that explicit.

> This spec also raises another open question about what happens if NaN is passed to a numeric argument but that argument is not used. Should the function throw or not?

In most cases I'd think it should throw. The usual pattern is that we coerce and validate all arguments before doing any processing. (https://github.com/tc39/how-we-work/pull/119 would write that rule down.)

Reply:
> Yes, imo extensive validation should be performed on all inputs even if they end up being unused.

Absolutely, yes. In general, input validation is done in full before any operation code is run, at least in WebIDL-defined things. It's a good thing.